File: MicroFrontend/appMicro/ios/App/App/plugins/DigitalWallet/utils/DigitalCardManagement.swift
import Foundation
import AntelopSDK
import Capacitor

struct DigitalCardStructure: Encodable {
    var id: String
    var status: String
    var bin: String
    var lastDigits: String
    var expirationDate: String
    var imageIsLoaded: Bool
}

class DigitalCardManagement: CAPPlugin {

    public static func mapStructure( key:String, value : DigitalCard) -> DigitalCardStructure {
        return DigitalCardStructure(
            id: value.getId(),
            status: DigitalCardManagement.convertDigitalCardStatusToString(value.getStatus()),
            bin: DigitalCardManagement.getBin(value.getCardInfo()),
            lastDigits: DigitalCardManagement.getLastDigits(value.getCardInfo()),
            expirationDate: DigitalCardManagement.getExpiryDate(value.getCardInfo()),
            imageIsLoaded: DigitalCardManagement.imageIsLoaded(value)
        )
    }

    private static func convertDigitalCardStatusToString(_ digitalCardServiceStatus: CardStatus?) -> String {
    switch digitalCardServiceStatus {
        case .activating:
            return "activating"
        case .activationRefused:
            return "activationRefused"
        case .activationRequired:
            return "activationRequired"
        case .active:
            return "active"
        case .locked:
            return "locked"
        case .termsAndConditionsValidationRequired:
            return "termsAndConditionsValidationRequired"
        default:
            NSLog("HCESDK:AVV: DigitalCardManagement (convertDigitalCardStatusToString) : \(digitalCardServiceStatus?.description)")
            return ""
        }
    }

    private static func getLastDigits(_ digitalCardCardInfo: CardInfo?) -> String {
        if let cardInfo = digitalCardCardInfo?.getLastDigits() {
            return cardInfo
        }
        return ""
    }

    private static func getBin(_ digitalCardCardInfo: CardInfo?) -> String {
        if let cardInfo = digitalCardCardInfo?.getBin() {
            return cardInfo
        }
        return ""
    }

    private static func getExpiryDate(_ digitalCardCardInfo: CardInfo?) -> String {
        if let expiryDate = digitalCardCardInfo?.getExpiryDate() {
            let dateFormatter = DateFormatter()
            dateFormatter.dateFormat = "yyyy-MM-dd HH:mm:ss"
            return dateFormatter.string(from: expiryDate)
        }
        return ""
    }

    private static func imageIsLoaded(_ digitalCard: DigitalCard) -> Bool {
        return digitalCard.getCustomCardDisplay() != nil
    }
}

File: MicroFrontend/appMicro/android/app/src/main/java/plugins/DigitalWallet/utils/DigitalCardManagement.java
package plugins.DigitalWallet.utils;

import android.util.Log;

import com.getcapacitor.Plugin;
import java.text.SimpleDateFormat;
import java.util.Date;
import java.util.Locale;
import java.util.Optional;

import fr.antelop.sdk.card.CardInfo;
import fr.antelop.sdk.digitalcard.DigitalCard;
import fr.antelop.sdk.card.CardStatus;
import plugins.DigitalWallet.models.DigitalCardStructure;

public class DigitalCardManagement extends Plugin {

    public static DigitalCardStructure mapStructure(String key, DigitalCard value) {
        return new DigitalCardStructure(
            value.getId(),
            convertDigitalCardStatusToString(value.getStatus()),
            getBin(value.getCardInfo()),
            getLastDigits(value.getCardInfo()),
            getExpiryDate(value.getCardInfo()),
            imageIsLoaded(value)
        );
    }

    private static String convertDigitalCardStatusToString(CardStatus status) {
        if (status == null) return "";
        switch (status) {
            case Activating: return "activating";
            case ActivationRefused: return "activationRefused";
            case ActivationRequired: return "activationRequired";
            case Active: return "active";
            case Locked: return "locked";
            case TermsAndConditionsValidationRequired: return "termsAndConditionsValidationRequired";
            default:
                Log.i("HCESDK:AVV", "DigitalCardManagement (convertDigitalCardStatusToString): " + status);
                return "";
        }
    }

    private static String getLastDigits(CardInfo cardInfo) {
        return Optional.ofNullable(cardInfo)
            .map(CardInfo::getLastDigits)
            .orElse("");
    }

    private static String getBin(CardInfo cardInfo) {
        return Optional.ofNullable(cardInfo)
            .map(CardInfo::getBin)
            .orElse("");
    }

    private static String getExpiryDate(CardInfo cardInfo) {
        Date expiry = Optional.ofNullable(cardInfo)
            .map(CardInfo::getExpiryDate)
            .orElse(null);

        if (expiry == null) return "";

        SimpleDateFormat formatter = new SimpleDateFormat("yyyy-MM-dd HH:mm:ss", Locale.US);
        return formatter.format(expiry);
    }

    private static boolean imageIsLoaded(DigitalCard card) {
        return true;
    }
}
